exchanges/orderbook/unsafe.go

```diff
@@ -0,0 +1,62 @@
+package orderbook
+
+import (
+	"sync"
+	"time"
+)
+
+// Unsafe is an exported linked list reference to current the current bid/ask
+// heads and a reference to the underlying depth mutex. This allows for the
+// exposure of the internal list to a strategy or subsystem. The bid and ask
+// fields point to the actual head fields contained on both linked list structs,
+// so that this struct can be reusable and not needed to be called on each
+// inspection.
+type Unsafe struct {
+	BidHead **Node
+	AskHead **Node
+	m       *sync.Mutex
+
+	// UpdatedViaREST defines if sync manager is updating this book via the REST
+	// protocol then this book is not considered live and cannot be trusted.
+	UpdatedViaREST *bool
+	LastUpdated    *time.Time
+	*Alert
+}
+
+// Lock locks down the underlying linked list which inhibits all pending updates
+// for strategy inspection.
+func (src *Unsafe) Lock() {
+	src.m.Lock()
+}
+
+// Unlock unlocks the underlying linked list after inspection by a strategy to
+// resume normal operations
+func (src *Unsafe) Unlock() {
+	src.m.Unlock()
+}
+
+// LockWith locks both books for the context of cross orderbook inspection.
+// WARNING: When inspecting diametrically opposed books a higher order mutex
+// MUST be used or a dead lock will occur.
+func (src *Unsafe) LockWith(dst sync.Locker) {
+	src.m.Lock()
+	dst.Lock()
+}
+
+// UnlockWith unlocks both books for the context of cross orderbook inspection
+func (src *Unsafe) UnlockWith(dst sync.Locker) {
+	dst.Unlock() // Unlock in reverse order
+	src.m.Unlock()
+}
```

Comment:
Just a thought on this dual lock system. You may wish to somehow protect dual locks from individual locks. As in, if you `LockWith`, you can also just call `Unlock` on each individually. You may be able to do it atomically to state that there is a dualLock in place, so cannot just call Unlock.
Since its for external use I'm less concerned, but I thought I'd just say the availability of both does present an opportunity for panics.

Reply:
Individual locks should not be an issue as you call lock first and then unlock. The problem occurs when you are locking multiple books for inspection. e.g. BTC-USD exch1, BTC-USDT exch2  in two routines we could check this set at the same time locking book1 and then book2, this is fine but if one routine locks it out of order say, locks book2 and then book1 then at the same time another routine locks book1 and then book2 that is where the dead lock occurs.

Reply:
Is there value in preventing a lock if either one is currently locked? By having an atomic switch first, whether dual or single lock, at least when you attempt to lock either of those pairs again, you can see the atomic status and return with the suggestion that the opportunity has been lost as a currency is already under a lock. Like, if you attempt to lock a lock, its blocking until it unlocks, but the atomic will allow you to check before you attempt.

Reply:
Sure but it might not needed on the GCT library you can easily add that externally for whatever you need:

```go
// Depth type to expand specific functionality for the orderbook.Unsafe type
type Depth struct {
	orderbook.Unsafe
        isLocked uint32 
}

func (d *Depth) IsLockedBeforeLock() error {
      // atomic fun 
}
```

Reply:
In that case above you would need to share a semaphore between relationships and those relationships would need to be tracked at a higher level as well. Structuring locking sequence might more efficient than introducing atomic checks.

Reply:
Yeah fair enough. Can be left for any theoretical external packages to manage 🎉